### PR TITLE
Disable the "Strip Linked Product" build setting when using the "arch…

### PR DESCRIPTION
**Pull request by abrindam**
Fix for #2280 and friends.

Disable the "Strip Linked Product" build setting (STRIP_INSTALLED_PRODUCT), when using the "archive" command to build.

This keeps the debug information around so a valid dSYM can be generated.

This change is necessary since Carthage moved to using the "archive" command to build. Most products have `STRIP_INSTALLED_PRODUCT` set to true, but this does not kick in unless you have `DEPLOYMENT_POSTPROCESSING` set to true as well, and that is false for most products. However, using "archive" instead of "build" magically overrides `DEPLOYMENT_POSTPROCESSING` to true, resulting in the product being stripped of debug symbols. As a result dSYM generation produces effectively no information for builds that are built with "archive" (e.g. OSX and iOS device, but not iOS simulator)

This also restores the 0.25 behavior that built frameworks are not stripped. Prior to fix, stripping was happening with the archive command, which may have caused additional issues that were not reported. With this change, we are back to the old behavior and stripping is done at framework copy time.

Testing done:
1. Build a framework using Carthage
2. Use "grep" to check the produced framework for a known source file name.
3. Use "grep" to check the produced dSYM for a known source file name
4. Run an app that uses this framework in Xcode. Open Debug->Debug Workflow->Shared Libraries and ensure the framework has "Debug Symbols" as "Yes"

Prior to fix, all checks above failed, and post fix, all checks above succeed.

### CODE REVIEW

**Comment by blender** [comment]
Hi,

please correct me if I am wrong but I don't see that frameworks will be stripped from debug symbols in the copy-framework phase.

To recap, the release builds of frameworks will now contain desymbolication information by default which might be not preferred by most.

**Comment by abrindam** [comment]
Hi @blender,

It appears you are correct, `carthage copy-frameworks` does not strip debug symbols. I was wrong. Therefore you are correct that in that case the release build will contain debug symbols.

However, I have tested with Carthage 0.25 and this behavior occurs there as well. So I would argue this is not a regression but a restoration to the way things used to work. Changing when debug symbols were stripped was not the intent of the changes that went in between 0.25 and 0.26 (the "archive" change, for short). So I feel it makes sense to put things back to the way they were and then move forward from there.

In other words, nobody complained about the debug symbols when they were included before. However, many are complaining about the invalids dSYMs now. So from a priority perspective I believe the cleanest thing to do is to revert to the 0.25 model.

Additionally, there were comments in the linked issue #2280 that indicated people *wanted* the built frameworks (in `Carthage/Build`) to contain debug information for their own convenience. So this argues for a strip-on-copy rather than strip-on-build approach.

If we want to fix the problem you're raising, I think it would be appropriate for `carthage copy-frameworks` to strip debug symbols based on the Xcode setting `COPY_PHASE_STRIP`. This setting affects the vanilla "Copy Files" phase in Xcode, and as `carthage copy-frameworks` is attempting to be "Copy Files On Steroids", it seems appropriate to honor that as well.

**Comment by blender** [comment]
You read my mind.


Would you be open to try to change copy-frameworks?

This way the consumer of the frameworks can decide whether to strip the symbols or not during copy. I imagine that in debug one would set `COPY_PHASE_STRIP` to no strip and on release to yes.

This would be an elegant way to solve the problem.


If you don't feel like changing this now I'm fine with restoring the old behavior.

**Comment by abrindam** [comment]
Let me give it a try. It's a bit more involved change that this current pull request, but I'm game.

BTW, from what I see `COPY_PHASE_STRIP` is on for release and off for debug by default in Xcode, just as you theorized.

**Comment by blender** [comment]
Thanks, that is very kind of you.

Just as a heads up including tests for the change would be a nice idea. I know other maintainers are going to look for them too when reviewing the PR ;)

**Comment by mdiep** [comment]
Does archiving the app strip the source code from the embedded frameworks?

**Comment by blender** [comment]
@mdiep I just tried building Alamofire with `carthage v0.28.0`, adding a copy-framework phase, archiving an app and it seems that it does strip the symbols. Nice catch.

How to verify: 

```
$otool -tv Carthage/Build/iOS/Alamofire.framework/Alamofire | grep Alamo # Brace for impact
...
$ otool -t "/Users/blender/Library/Developer/Xcode/Archives/2018-02-08/TestsApp 08.02.18, 23.02.xcarchive/Products/Applications/TestsApp.app/Frameworks/Alamofire.framework/Alamofire" | grep Alamo
# no trace
```
So seems that there isn't additional work here @abrindam.

**Comment by abrindam** [comment]
OK, great. Hopefully we can move forward with this pull request then.

I actually coded something up already that seems to work, for symmetry with the standard Xcode copy files we might want to include it anyways. I'll open a separate issue/PR however given this is no longer critical.

**Comment by blender** [comment]
@mdiep  @abrindam 

Hello, sorry to break the news but I just tried with the latest master and archiving the app does not strip the symbols from embedded frameworks.

I got mislead by the fact that before this merge there were no symbols at all for the arm slice so when I checked the embedded library I saw no symbols because it was `lipoed` from the x84 arch which was the only slice with symbols to begin with.

So long story short, the behavior is now restored to `v0.25` but imho the `copy-framework` changes are still needed and relevant.

**Comment by abrindam** [comment]
OK @blender I'll open a separate request with my changes to strip on copy. I'm unfortunately about to embark on a 21 hour flight to India so it'll be a bit before I can get it posted...

**Comment by blender** [comment]
Have a safe flight. No worries.

**Comment by blender** [comment]
@abrindam ping

**Comment by abrindam** [comment]
Hey folks, apologies for totally disappearing. Been a crazy couple weeks. 

I've been soaking my changes by using my local copy of Carthage and I haven't seen any issues. Will open a pull request momentarily for my changes.
